<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>MCAST - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>

    <div class="prog_logo">
      <img src="images/mcast_icon.png" alt="MCAST Logo">
      <h1>MCAST</h1>
      <h2>Motif Cluster Alignment and Search Tool</h2>
    </div>
    
    <div class="usage">
      <h3>Usage:</h3>
      <code>mcast [options] &lt;motifs&gt; &lt;sequence database&gt;</code>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <p>MCAST searches a sequence database for statistically significant
      clusters of non-overlapping occurrences of a given set of motifs.</p>

      <p>A motif "hit" is a sequence position that is sufficiently similar to
      a motif in the query, where the score for a motif at a particular
      sequence position is computed without gaps. To compute the
      <i>p-</i>value of a motif score, MCAST assumes that the sequences in the
      database were generated by a 0-order Markov process (see option
      <span class="popt">--bgfile</span>, below). To be considered a hit, the
      <i>p</i>-value of the motif alignment score must be less than the
      significance threshold, <span class="pdat">pthresh</span> (see option
      <span class="popt">--motif-pthresh</span>, below). Note that MCAST
      searches for hits on both strands of the sequences.</p>

      <p>A cluster of non-overlapping hits is called a "match". The user
      specifies the maximum allowed distance between the hits in a match using
      the <span class="popt">--max-gap</span> option. Two hits separated by
      more than the maximum allowed gap will be reported in separate
      matches.</p>

      <p style="margin-bottom:0">The <i>p-</i>value of a hit is converted to
      a "p-score" in order to compute the total score of the match it
      participates in. The p-score for a hit with <i>p-</i>value
      <span class="pdat">p</span> is</p>
      <div class="indent">
        S = -log<sub>2</sub>(<span class="pdat">p</span>/<span class="pdat">pthresh</span>),
      </div>
      <p style="margin-top:0">The total score of a match is the sum of the
      p-scores of the hits making up the match.</p>

      <p>MCAST searches for all possible matches between the query motifs and
      the sequences in the database, and reports the matches with the largest
      scores in decreasing order. Three types of statistical confidence
      estimates (<i>p</i>-value, <i>E</i>-value, and <i>q</i>-value) are
      estimated for each score, and the reported matches can be filtered by
      applying <i>p</i>-value or <i>q</i>-value thresholds (see the options
      <span class="popt">--output-pthresh</span> and
      <span class="popt">--output-qthresh</span> below).</p>

      <p>In order for MCAST to compute statistical confidence estimates, at
      least 100 matches must be found. If the database contains too few
      sequences, or if certain other options are made too stringent, then too
      few matches may exist for significance statistics to be computed. In
      this case, the <i>p</i>-value, <i>q</i>-value, and <i>E</i>-value
      columns are set to "NaN", and all matches are printed. This limitation
      can be overcome by specifying the <span class="popt">--synth</span> and
      <span class="popt">---bgfile</span> options. When those options are set,
      synthetic sequences will be generated from the provided background model
      and used to estimate significance statistics.</p>

      <p>When computing statistical confidence estimates, MCAST must retain
      the matches in memory until the final distribution of scores can be
      estimated. This means that the scanning of genome sized datasets has the
      potential to exhaust all available memory. To avoid this problem, MCAST
      uses reservoir sampling of the match scores, and limits the number of
      matches that are kept in memory. The default number of matches kept in
      memory is 100,000, but this value can be adjusted via the
      <span class="popt">--max-stored-scores</span> option. If the maximum
      number of stored matches is reached, then MCAST will drop the least
      significant half of the matches. This behavior may result in matches
      missing from the MCAST output, even though they would have satisified
      the user-specified <i>p</i>-value or <i>q</i>-value threshold.</p>

      <p style="margin-bottom:0">A full description of the algorithm may be found in:</p>
      <div class="citation">
        <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/19/suppl_2/ii16">
        Bailey and Noble. "Searching for statistically significant regulatory
        modules." <i>Bioinformatics (Proceedings of the European Conference on
        Computational Biology)</i>. 19(Suppl. 2):ii16-ii25, 2003.</a>
      </div>
    </div>

    <h3>Input</h3>
    <div class="indent">
      <h4>Motifs</h4>
      <div class="indent">
        <p>A file containing <a href="meme-format.html">MEME formatted motifs</a>.
        Outputs from MEME and DREME are supported along with minimal MEME
        format for which there are conversion scripts available to support
        other formats. Input motifs that are likely to appear in the
        sequences.</p>
      </div>

      <h4>Sequence Database</h4>
      <div class="indent">
        <p>A collection of DNA sequences in
        <a href="fasta-format.html">FASTA</a> format.</p>
      </div>
    </div>

    <h3>Output</h3>
    <div class="indent">
      <p style="margin-bottom:0">MCAST will create a directory named
      <code>mcast_out</code> (the name of this directory can be overridden via the
      <span class="popt">--o</span> or <span class="popt">--oc</span> options)
      The directory will contain:</p>
      <ul>
        <li>A file named <code>mcast.xml</code> describing the inputs to MCAST
        in XML format</li>
        <li>A file named <code>cisml.xml</code> reporting the matches in XML
        format using the <a href="http://zlab.bu.edu/~phaverty/SupplementalData/CisML">CisML</a>
        schema</li>
        <li>A file named <a href="#html_out"><code>mcast.html</code></a>
        reporting the matches in HTML format</li>
        <li>A file named <a href="#txt_out"><code>mcast.txt</code></a>
        reporting the matches in tab-delimited format</li>
        <li>A file named <code>mcast.gff</code> reporting the matches in
        <a href="http://www.sanger.ac.uk/resources/software/gff/">GFF format</a></li>
        <li>A file named <a href="#wig_out"><code>mcast.wig</code></a>
        reporting the matches in
        <a href="https://cgwb.nci.nih.gov/goldenPath/help/wiggle.html">wiggle track format</a></li>
      </ul>
    </div>
    <h3>Options</h3>
    <div class="indent">
      <table class="options">
        <thead>
          <tr>
            <th>Option</th>
            <th>Parameter</th>
            <th>Description</th>
            <th>Default Behaviour</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th colspan="4">General Options</th>
          </tr>
          <tr>
            <td class="popt">-bgfile</td><td><span class="pdat">background file</span></td>
            <td>A file with background frequencies of letters. This expects a
            <a href="bfile-format.html">markov background model</a>.
            This also accepts special values like --nrdb-- which has the same
            effect as not supplying a background, --uniform-- which uses a
            uniform background and --motif-- which uses the background
            specified in the motif file.</td>
            <td>Uses frequencies embedded in the application from the
            non-redundant database.</td>
          </tr>
          <tr>
            <td class="popt">--bgweight</td><td><span class="pdat">weight</span></td>
            <td>Add <span class="pdat">weight</span> times the background
            frequency to the corresponding letter counts in each motif when 
            converting them to postion specific scoring matrices.</td>
            <td>A weight of 4.0 is used.</td>
          </tr>
          <tr>
            <td class="popt">--max-gap</td><td><span class="pdat">max gap</span></td>
            <td>The value of <span class="pdat">max gap</span> specifies the
            longest distance allowed between two hits in a match. Hits
            separated by more than <span class="pdat">max gap</span> will be
            placed in different matches.
            <b>Note:</b> Large values of <span class="pdat">max gap</span>
            combined with large values of <span class="pdat">pthresh</span>
            may prevent MCAST from computing <i>E</i>-values.</td>
            <td>The maximum gap is set to 50.</td>
          </tr>
          <tr>
            <td class="popt">--max-stored-scores</td><td><span class="pdat">max</span></td>
            <td>Set the maximum number of scores that will be stored.
            Keeping a complete list of scores may exceed available memory.
            Once the number of stored scores reaches the maximum allowed,
            the least significant 50% of scores will be dropped.
            In this case, the list of reported motifs may be incomplete and 
            the q-value calculation will be approximate.</td>
            <td>The maximum number of stored matches is 100,000.</td>
          </tr>
          <tr>
            <td class="popt">--motif-pthresh</td><td><span class="pdat">pthresh</span></td>
            <td>sets the scale for calculating pscores for motif hits. The
            p-score for a hit with p-value <span class="pdat">p</span> is
            <div class="indent">
              S = -log<sub>2</sub>(<span class="pdat">p</span>/<span class="pdat">pthresh</span>),
            </div>
            </td>
            <td>The motif scaling pvalue is set to 0.0005.</td>
          </tr>
          <script>
            wropt("all-o", {"ddash": true, "dir": "mcast_out"});
            wropt("all-oc", {"ddash": true, "dir": "mcast_out"});
          </script>
          <tr>
            <td class="popt">--output-ethresh</td><td><span class="pdat">out <i>E</i>-value</span></td>
            <td>The <i>E</i>-value threshold for displaying search results. If
            the <i>E</I>-value of a match is greater than this value, then the
            match will not be printed. If any combination of
            <span class="popt">--output-ethresh</span>,
            <span class="popt">--output-pthresh</span>, or
            <span class="popt">--output-qthresh</span> is given, whichever
            option occurs last on the command-line will determine the
            effective output filter. </td>
            <td>The <i>E</i>-value threshold is 10.0.</td>
          </tr>
          <tr>
            <td class="popt">--output-pthresh</td><td><span class="pdat">out <i>p</i>-value</span></td>
            <td>The <i>p</i>-value threshold for displaying search results. If
            the <i>p</i>-value of a match is greater than this value, then the
            match will not be printed. If any combination of
            <span class="popt">--output-ethresh</span>,
            <span class="popt">--output-pthresh</span>, or
            <span class="popt">--output-qthresh</span> is given, whichever
            option occurs last on the command-line will determine the
            effective output filter.</td>
            <td>The <i>E</i>-value is used as the threshold. See 
            <span class="popt">--output-ethresh</span> option.</td>
          </tr>
          <tr>
            <td class="popt">--output-qthresh</td><td><span class="pdat">out <i>q</i>-value</span></td>
            <td>The <i>q</i>-value threshold for displaying search results. If
            the <i>q</i>-value of a match is greater than this value, then the
            match will not be printed. If any combination of
            <span class="popt">--output-ethresh</span>,
            <span class="popt">--output-pthresh</span>, or
            <span class="popt">--output-qthresh</span> is given, whichever
            option occurs last on the command-line will determine the
            effective output filter.</td>
            <td>The <i>E</i>-value is used as the threshold. See 
            <span class="popt">--output-ethresh</span> option.</td>
          </tr>
          <tr>
            <td class="popt">--synth</td><td></td>
            <td>Use synthetic scores for distribution.</td>
            <td></td>
          </tr>
          <script>wropt("all-verbosity");</script>
        </tbody>
      </table>
    </div>

    <h3 id="html_out">HTML output</h3>
    <div class="indent">
      <p style="margin-bottom:0">The HTML output contains</p>
      <ul>
        <li>A list of the motifs, and the best possible "hit" for each</li>
        <li>A list of matches. Each match record contains
        <ul>
          <li>The name of the sequence</li> 
          <li>The starting and ending coordinates for the match. The
          coordinates are 1-based, and start at the first position in the
          sequence.</li>
          <li>The match score</li>
          <li>The match p-value</li>
          <li>The match e-value</li>
          <li>The match q-value</li>
          <li>A block diagram showing the relative positions of each motif
          "hit" in the match</li>
          <li>A detailed view of each match is available as a pop-up window.
          The detailed view shows the full sequence for the match,
          the alignment to the motifs, and the <i>p</i>-values for the motif
          hits</li>
        </ul></li>
        <li>The inputs to MCAST</li>
        <li>Text describing the MCAST results</li>
      </ul>
    </div>

    <h3 id="txt_out">Text output</h3>
    <div class="indent">
      <p style="margin-bottom:0">The plain text output contains a line for
      each match. Each line contains the following fields:</p>
      <ul style="margin-bottom:5px">
        <li>An id string for the match</li>
        <li>The name of the sequence containing the match</li>
        <li>The starting and ending coordinates of the match (1-based)</li>
        <li>The match score</li>
        <li>The match p-value</li>
        <li>The match e-value</li>
        <li>The match q-value</li>
        <li>The sequence of the matched region</li>
      </ul>
      <p style="margin-top:0">The lines are sorted by score in descending 
      order.</p>
    </div>

    <h3 id="wig_out">Wiggle output</h3>
    <div class="indent">
      <p style="margin-bottom:0">The wiggle track output contains the
      following entries:</p>
      <ul style="margin-bottom:5px">
        <li>A track line containing:
        <ul>
          <li>The track type</li>
          <li>The source of the track (MCAST)</li>
        </ul></li>
        <li>A step size line containing:
        <ul>
          <li>The sequence name</li>
          <li>The width of the match</li>
        </ul></li>
        <li>A data line containing:
        <ul>
          <li>The start position of the match (closed, 1-based coordinates)</li>
          <li>A score based on the p-value of the match: -log(<i>p</i>-value) where</li>
        </ul></li>
      </ul>
    <p style="margin-top:0">The wiggle track output is sorted by sequence name and position.</p>
    </div>
    <h3>Citing</h3>
    <div class="indent">
      <p>
        If you use MCAST in your research please cite the following paper:<br />
        <span class="citation">
          Timothy Bailey and William Stafford Noble, 
          &quot;Searching for statistically significant regulatory modules&quot;, 
          <i>Bioinformatics (Proceedings of the European Conference on Computational Biology)</i>,
          19(Suppl. 2):ii16-ii25, 2003.
          <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/19/suppl_2/ii16">[full text]</a>
        </span>
      </p>
    </div>

  </body>
</html>

