<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Overview - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <style>
      img.central {
        display:block;
        margin-left:auto;
        margin-right:auto;
      }
      table.overview th {
        text-align: left;
      }
      table.overview > tbody + tbody  th {
        padding-top: 40px;
      }
      table.overview tr:hover {
        background-color: #E4ECEC;
      }
      table.overview tr:nth-child(1):hover {
        background-color: inherit;
      }

    </style>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <p>The MEME Suite allows the biologist to discover novel motifs in
    collections of unaligned nucleotide or protein sequences and to search
    for motif occurrences in sequence databases.</p>
    
    <p>The MEME Suite is comprised of a collection of tools that work
    together, as shown below. Not all the tools are available as
    webservices, so to get the full power of the MEME Suite you will need to  <!-- ../meme-download.html -->
    <a href="download.html">download</a> and <a href="install.html">install</a>
    a local copy of the software. To see what has changed recently you can
    peruse the <a href="release-notes.html">release notes</a>.</p>

    <map name="webserver_schema_map">
      <area shape="rect" coords="133,72,204,84" href="meme.html" alt="MEME"/>
      <area shape="rect" coords="133,85,204,99" href="dreme.html" alt="DREME"/>
      <area shape="rect" coords="133,100,204,111" href="meme-chip.html" alt="MEME-ChIP"/>
      <area shape="rect" coords="133,112,204,125" href="glam2.html" alt="GLAM2"/>
      <area shape="rect" coords="370,51,442,75" href="tomtom.html" alt="Tomtom"/>
      <area shape="rect" coords="370,94,442,105" href="centrimo.html" alt="CentriMo"/>
      <area shape="rect" coords="370,106,442,119" href="ame.html" alt="AME"/>
      <area shape="rect" coords="370,120,442,134" href="spamo.html" alt="SpaMo"/>
      <area shape="rect" coords="370,156,442,169" href="fimo.html" alt="FIMO"/>
      <area shape="rect" coords="370,170,442,183" href="mast.html" alt="MAST"/>
      <area shape="rect" coords="370,184,442,196" href="mcast.html" alt="MCAST"/>
      <area shape="rect" coords="370,197,442,210" href="glam2scan.html" alt="GLAM2Scan"/>
      <area shape="rect" coords="370,236,442,257" href="gomo.html" alt="GOMo"/>
      <area shape="rect" coords="297,4,351,48" href="meme-format.html" alt="Motif Databases"/>
      <area shape="rect" coords="297,184,350,227" href="fasta-format.html" alt="Sequence Databases"/>
      <area shape="rect" coords="297,262,350,307" href="http://www.geneontology.org/"
      alt="Gene Ontology Terms Database"/>
      <area shape="rect" coords="8,152,264,164" href="meme.html" alt="MEME"/>
      <area shape="rect" coords="8,165,264,175" href="dreme.html" alt="DREME"/>
      <area shape="rect" coords="8,176,264,187" href="meme-chip.html" alt="MEME-ChIP"/>
      <area shape="rect" coords="8,188,264,198" href="glam2.html" alt="GLAM2"/>
      <area shape="rect" coords="8,199,264,209" href="tomtom.html" alt="TOMTOM"/>
      <area shape="rect" coords="8,210,264,221" href="centrimo.html" alt="CentriMo"/>
      <area shape="rect" coords="8,222,264,232" href="ame.html" alt="AME"/>
      <area shape="rect" coords="8,233,264,244" href="spamo.html" alt="SpaMo"/>
      <area shape="rect" coords="8,245,264,255" href="fimo.html" alt="FIMO"/>
      <area shape="rect" coords="8,256,264,266" href="mast.html" alt="MAST"/>
      <area shape="rect" coords="8,267,264,278" href="mcast.html" alt="MCAST"/>
      <area shape="rect" coords="8,279,264,290" href="glam2scan.html" alt="GLAM2Scan"/>
      <area shape="rect" coords="8,291,264,302" href="gomo.html" alt="GOMo"/>
    </map>
    <img src="images/webserver_schema.png" class="central" usemap="#webserver_schema_map">

    <table class="overview">
      <tbody>
      <tr>
        <th colspan="3">Motif Discovery</th>
      </tr>

      <tr>
        <td><a href="meme.html">MEME</a></td>
        <td>Find ungapped motifs in unaligned DNA, RNA or protein sequences.</td>
        <td><a href="examples/meme_example_output_files/meme.html">Sample</a></td>
      </tr>

      <tr>
        <td><a href="meme-chip.html">MEME ChIP</a></td>
        <td>Comprehensive motif analysis of large nucleotide datasets such as
        those from ChIP-seq or CLIP-seq experiments.</td>
        <td>
          <a href="examples/memechip_example_output_files/index.html">Sample</a>
        </td>
      </tr>

      <tr>
        <td><a href="dreme.html">DREME</a></td>
        <td>Find short, ungapped motifs in large sets of DNA or RNA sequences.  Also allows
        discriminative motif analysis using a set of control sequences.</td>
        <td>
          <a href="examples/dreme_example_output_files/dreme.html">Sample</a>
        </td>
      </tr>

      <tr>
        <td><a href="glam2.html">Glam2</a></td>
        <td>Find <b>gapped</b> motifs in DNA or protein sequences.
          It has a <a href="glam2_tut.html">tutorial</a>.</td>
        <td>
          <a href="examples/glam2_example_output_files/glam2.html">Sample</a>
        </td>
      </tr>
      </tbody>

      <tbody>
      <tr>
        <th colspan="3">Motif Enrichment Analysis</th>
      </tr>

      <tr>
        <td><a href="ame.html">AME</a></td>
        <td>Motif Enrichment Analysis: find known DNA or RNA motifs 
        that are relatively enriched in the input sequences compared to
        shuffled version of those sequences or control sequences.</td>
        <td>&nbsp;</td>
      </tr>

      <tr>
        <td><a href="centrimo.html">CentriMo</a></td>
        <td>Local Motif Enrichment Analysis: Find motifs that are enriched
        in local regions in equal-length sequences.</td>
        <td>
          <a href="examples/centrimo_example_output_files/centrimo.html">Sample</a>
        </td>
      </tr>

      <tr>
        <td><a href="spamo.html">SpaMo</a></td>
        <td>Motif Spacing Analysis: Find known motifs that occur with preferred
        spacings relative to a primary motif in a set of DNA sequences.</td>
        <td>
          <a href="examples/spamo_example_output_files/spamo.html">Sample</a>
        </td>
      </tr>

      <tr>
        <td><a href="gomo.html">GOMO</a></td>
        <td>Genome Ontology Motif Enrichment: Identify possible roles (Gene Ontology terms) 
          for DNA binding motifs.</td>
        <td>&nbsp;</td>
      </tr>
      </tbody>


      <tbody>
      <tr>
        <th colspan="3">Motif Search</th>
      </tr>

      <!--
      <tr>
        <td><a href="beadstring.html">Beadstring</a></td>
        <td>
          Searches a sequence database for a given collection of motifs.  The
          motifs are represented using a hidden Markov model with a linear
          topology, thus requiring that the motifs appear in a specified
          order.
        </td>
        <td>Sample</td>
      </tr>
      -->

      <tr>
        <td><a href="fimo.html">FIMO</a></td>
        <td>
          Search a sequence database for occurrences of known motifs.  This
          program treats each motif independently and reports all putative motif
          occurrences below a specified <i>p</i>-value threshold.
        </td>
        <td><a href="examples/fimo_example_output_files/fimo.html">Sample</a></td>
      </tr>

      <tr>
        <td><a href="mast.html">MAST</a></td>
        <td>
          Search a sequence database for occurences of known motifs.  This
          program assumes exactly one occurrence of each motif per sequence, and
          each sequence in the database is assigned a <i>p</i>-value, based on
          the product of the <i>p</i>-values of the individual motif occurrences in
          that sequence.
        </td>
        <td><a href="examples/mast_example_output_files/mast.html">Sample</a></td>
      </tr>

      <tr>
        <td><a href="mcast.html">MCAST</a></td>
        <td>
          Search a sequence database for <b>clusters</b> of known motifs.
          <!--Like beadstring and miao,--> 
          mcast employs a motif-based hidden
          Markov model, using a star topology and a novel scoring algorithm.
          The motifs may appear in any order.
        </td>
        <td><a href="examples/mcast_example_output_files/mcast.html">Sample</a></td>
      </tr>

      <tr>
        <td><a href="glam2scan.html">Glam2Scan</a></td>
        <td>Search for occurences of gapped motifs, discovered by GLAM2.</td>
        <td><a href="examples/glam2scan_example_output_files/glam2scan.html">Sample</a></td>
      </tr>
    </tbody>
    <tbody>

      <tr>
        <th colspan="3">Motif Comparison</th>
      </tr>
      <tr>
        <td><a href="tomtom.html">Tomtom</a></td>
        <td>Find motifs that are similar to a given DNA or RNA motif by searching a database of 
        known motifs.</td>
        <td><a href="examples/tomtom_example_output_files/tomtom.html">Sample</a></td>
      </tr>
    </tbody>
    <tbody>

      <tr>
        <th colspan="3">Additional Primary Tools</th>
      </tr>

      <tr>
        <td><a href="ama.html">AMA</a></td>
        <td>Print the Average Motif Affinity score of each sequence in a database.  
        The score is calculated by averaging the likelihood ratio scores for all
        feasible binding events to the given sequence and to its reverse strand. 
        </td>
      </tr>
    </tbody>

    <tbody>
      <tr id="motif_conversion_utilities">
        <th colspan="2">Motif Format Conversion Scripts</th>
        <td rowspan="14"><a href="motif_conversion.html">Foreign Motif Formats</a></td>
      </tr>
        <tr>
          <td><a href="beeml2meme.html">beeml2meme</a></td>
          <td>Convert an BEEML matrix file to MEME format.</td>
        </tr>
        <tr>
          <td><a href="chen2meme.html">chen2meme</a></td>                  
          <td>Convert a CHEN matrix file to MEME format.</td>
        </tr>
        <tr>
          <td><a href="iupac2meme.html">iupac2meme</a></td>
          <td>Convert an IUPAC string to MEME format.</td>
        </tr>
        <tr>
          <td><a href="jaspar2meme.html">jaspar2meme</a></td>
          <td>Convert a directory of JASPAR files to MEME format.</td>
        </tr>
        <tr>
          <td><a href="meme2meme.html">meme2meme</a></td>
          <td>Convert and merge multiple MEME formatted files.</td>
        </tr>
        <tr>
          <td><a href="nmica2meme.html">nmica2meme</a></td>
          <td>Convert a nestedMICA (BioTiffin/XMS) matrix file to MEME format.</td>
        </tr>
        <tr>
          <td><a href="priority2meme.html">priority2meme</a></td>
          <td>Convert a PRIORITY matrix file to MEME format.</td>
        </tr>
        <tr>
          <td><a href="rna2meme.html">rna2meme</a></td>
          <td>Convert a FASTA file with short RNA sequences into motifs for DNA
            they might bind in MEME format.</td>
        </tr>
        <tr>
          <td><a href="scpd2meme.html">scpd2meme</a></td>
          <td>Convert an SCPD matrix file to MEME format.</td>
        </tr>
        <tr>
          <td><a href="sites2meme.html">sites2meme</a></td>
          <td>Convert files containing sites into MEME format.</td>
        </tr>
        <tr>
          <td><a href="taipale2meme.html">taipale2meme</a></td>
          <td>Convert a tab-separated file exported from a spreadsheet of
            Taipale results to MEME format.</td>
        </tr>
        <tr>
          <td><a href="tamo2meme.html">tamo2meme</a></td>
          <td>Convert a TAMO matrix file to MEME format.</td>
        </tr>
        <tr>
          <td><a href="transfac2meme.html">transfac2meme</a></td>                  
          <td>Convert a TRANSFAC matrix file to MEME format.</td>
        </tr>
        <tr>
          <td><a href="uniprobe2meme.html">uniprobe2meme</a></td>                  
          <td>Convert a UNIPROBE matrix file to MEME format.</td>
        </tr>

    </tbody>


    <tbody>

      <tr id="conversion_utilities">
        <th colspan="3">File Format Conversion Utilities</th>
      </tr>



        <tr>
          <td><a href="clustalw2fasta.html">clustalw2fasta</a></td>
          <td>Convert a Clustalw multiple alignment into FASTA format.</td>
        </tr>

        <tr>
          <td><a href="clustalw2phylip.html">clustalw2phylip</a></td>
          <td>Convert a Clustalw multiple alignment into Phylip format.</td>
        </tr>

        <!-- Not currently supported
        <tr>
          <td><a href="draw-mhmm.html">draw-mhmm</a></td>
          <td>
            Convert a motif-based HMM into a format suitable for drawing by 
            <a href="http://www.research.att.com/sw/tools/graphviz">graphviz</a>.
          </td>
        </tr>
        -->

        <tr>
          <td><a href="glam2format.html">glam2format</a></td>
          <td>Convert glam2 motifs to standard alignment formats.</td>
        </tr>

        <tr>
          <td><a href="godag-format.html#obo2dag">obo2dag</a></td>
          <td>Convert a Gene Ontology OBO file into a GO DAG file.</td>
        </tr>



        <!-- Not currently supported
        <tr>
          <td><a href="log-hmm.html">log-hmm</a></td>
          <td>Convert a motif-based HMM from probabilities to log probabilities.</td>
        </tr>
        -->

        <tr>
          <td><a href="mast2txt.html">mast2txt</a></td>
          <td>Convert MAST XML output into plain text.</td>
        </tr>


        <tr>
          <td><a href="readseq.html">readseq</a></td>
          <td>Format conversion utility for sequence data.</td>
        </tr>

      </tbody>
      <tbody>

      <tr id="assorted_utilities"></a>
        <th colspan="3">Other Utilities</th>
      </tr>

        <tr>
          <td><a href="alphtype.html">alphtype</a></td>
          <td>
            Classify a string passed as a command line argument
            as an instance of the DNA or protein alphabet.
          </td>
        </tr>

        <tr>
          <td><a href="ama-qvalues.html">ama-qvalues</a></td>
          <td>
            Add q-values to AMA output.
          </td>
        </tr>

        <tr>
          <td><a href="ceqlogo.html">ceqlogo</a></td>
          <td>Create motif logos.</td>
        </tr>

        <tr>
          <td><a href="compute-prior-dist.html">compute-prior-dist</a></td>
          <td>Compute the distribution of priors in a 
          <a href="meme.html#MEME-PSP">MEME PSP format file</a>.</td>
          <td><a
          href="examples/compute_prior_dist_example_output_files/prior.dist.txt">Sample</a></td>
        </tr>

        <tr>
          <td><a href="fasta-get-markov.html">fasta-get-markov</a></td>
          <td>Estimate a Markov model from a FASTA file of sequences.</td>
        </tr>

        <tr>
          <td><a href="fasta-io.html">fasta-io</a></td>
          <td>Read and write FASTA files.</td>
        </tr>

        <tr>
          <td><a href="fasta-subsample.html">fasta-subsample</a></td>
          <td>Extract a random selection of the sequences in a FASTA file. Can also subsample the sequences themselves.</td>
        </tr>

        <tr>
          <td><a href="fitevd.html">fitevd</a></td>
          <td>Fit an extreme value distribution to data.</td>
        </tr>

        <tr>
          <td><a href="gendb.html">gendb</a></td>
          <td>Generate sequences from a Markov model.</td>
        </tr>

        <tr>
          <td><a href="getsize.html">getsize</a></td>
          <td>Print statistics about sequences read from a FASTA file.</td>
        </tr>

        <tr>
          <td><a href="glam2mask.html">glam2mask</a></td>
          <td>                    
            Mask glam2 motifs out of sequences, so that weaker motifs can be
            found.
          </td>
        </tr>

        <tr>
          <td><a href="gomo_highlight.html">gomo_highlight</a></td>                  
          <td>
            Identify GO terms which are implied by other GO terms, allowing the
            most specific GO terms to be highlighted in the conversion to html.
          </td>
        </tr>

        <tr>
          <td><a href="meme-io.html">meme-io</a></td>
          <td>Print summary of MEME HTML file to standard output as plain text.</td>
        </tr>

        <tr>
          <td><a href="meme-rename.html">meme-rename</a></td>
          <td>Easily rename MEME Suite HTML files to unique names incorporating
          the path name (rather than "meme.html").</td>
        </tr>

        <!-- Not currently supported
        <tr>
          <td><a href="mhmm.html">mhmm</a></td>
          <td>
            Create motif-based hidden Markov models (HMMs) of families of related 
            biosequences. The program takes as input a set of DNA or protein motif models constructed 
            by MEME and produces as output a single HMM containing the given motifs.
          </td>
        </tr>

        <tr>
          <td><a href="mhmme.html">mhmme</a></td>
          <td>
            Randomly generate sequences according to a given hidden Markov
            model.
          </td>
        </tr>

        <tr>
          <td><a href="mhmms.html">mhmms</a></td>
          <td>
            Search a sequence database using a Meta-MEME motif-based hidden 
            Markov model (HMM) of the kind produced by mhmm.
          </td>
        </tr>

        <tr>
          <td><a href="mhmmscan.html">mhmmscan</a></td>
          <td>
            Search a sequence database using a Meta-MEME motif-based hidden 
            Markov model (HMM) of the kind produced by mhmm. This program is similar to mhmms, 
            except that mhmmscan searches arbitrarily long sequences, and
            allows the model to match a given sequence multiple times, 
            and each match is reported separately.
          </td>
        </tr>
        -->

        <tr>
          <td><a href="pmp_bf.html">pmb_bf</a></td>
          <td>Calculate the statistical power of phylogentic motif models.</td>
        </tr>

        <tr>
          <td><a href="psp-gen.html">psp-gen</a></td>
          <td>
          Generate position-specific priors from positive (likely to
          contain a feature of interest) and negative (unlikely
          to contain a feature of interest) sequences for use as an additional input to
          MEME.
          </td>
        </tr>

        <tr>
          <td><a href="purge.html">purge</a></td>
          <td>Remove highly similar members of a set of sequences.</td>
        </tr>

        <tr>
          <td><a href="qvalue.html">qvalue</a></td>
          <td>Compute q-values from <i>p</i>-values.</td>
        </tr>

        <tr>
          <td><a href="reconcile-tree-alignment.html">reconcile-tree-alignment</a></td>
          <td>
            Given a tree and an alignment, identify the intersection of the sets
            of sequence IDs and leaf labels.  Trim the extra sequences and leaves
            and print the resulting alignment and tree.
          </td>
        </tr>

        <tr>
          <td><a href="reduce-alignment.html">reduce-alignment</a></td>
          <td>Extract specified columns from a multiple alignment.</td>
        </tr>

        <tr>
          <td><a href="remove-alignment-gaps.html">remove-alignment-gaps</a></td>
          <td>
            Remove from an alignment all columns that correspond to a gap in a
            specified species.
          </td>
        </tr>

        <!-- Not currently supported
        <tr>
          <td><a href="select-star-species.html">select-star-species</a></td>
          <td>Tries to find the subset of species whose subtree is "closet" to a start topology.</td>
        </tr>
                    -->

        <tr>
          <td><a href="shadow.html">shadow</a></td>
          <td>
            Perform phylogenetic shadowing on a given DNA alignment, using a given
            tree.
          </td>
        </tr>
      </tbody>

      <tbody>
        <tr><th colspan="3" id="formats">File Formats</th></tr>
        <tr>
          <td><a href="meme-format.html">MEME Motif</a></td>
          <td>The motif format which is supported by the MEME Suite.</td>
        </tr>
        <tr>
          <td><a href="bfile-format.html">Background Model</a></td>
          <td>Background frequencies for DNA or protein sequences.</td>
        </tr>
        <tr>
          <td><a href="fasta-format.html">FASTA Sequence</a></td>
          <td>DNA or protein sequences.</td>
        </tr>
        <tr>
          <td><a href="clustalw-format.html">ClustalW Alignment</a></td>
          <td>A multiple alignment of DNA or protein sequences.</td>
        </tr>
        <tr>
          <td><a href="transfac-format.html">TRANSFAC Motif</a></td>
          <td>A motif format which can be converted into MEME motifs.</td>
        </tr>
        <tr>
          <td><a href="glam2_alphabet.html">GLAM2 Alphabet</a></td>
          <td>A custom sequence alphabet for GLAM2. This can be used
            to provide alternate alphabets other than the standard DNA and
            protein.</td>
        </tr>
        <tr>
          <td><a href="godag-format.html">GO DAG</a></td>
          <td>A file format which stores the structure of the Gene Ontology so
            it can be used to improve GOMO output.</td>
        </tr>
      </tbody>

      <tbody>
        <tr><th colspan="3">Guides and Tutorials</th></tr>
        <tr>
          <td><a href="scripting.html">Script Access</a></td>
          <td>How to access MEME Suite web services from Perl or Python scripts.</td>
        </tr>
        <tr>
          <td><a href="install.html">Installation</a></td>
          <td>How to install a local copy of the MEME Suite.</td>
        </tr>
        <tr>
          <td><a href="release-notes.html">Release notes</a></td>
          <td>A list of changes included in the latest release.</td>
        </tr>
      </tbody>
      </table>

      <p>Visit the <a href="http://meme.nbcr.net">MEME Suite home page</a>.</p>

    <hr/>
    <h3 style="font-weight: normal;">
      Maintenance and development of the MEME Suite is funded by the 
      National Center for Research Resources grant NIH/NCRR R01 RR021692.
      The MEME Suite web server is funded by the 
      <a href= "http://www.sdsc.edu/NBCR">National Biomedical Computation Resource</a>.
    </h3>
    <hr/>
    <b><font>Developed and maintained by:</font></b>
    <table align="center">
      <tr>
        <td>
          <a href="http://www.uq.edu.au">
            <img src="images/uq_logo.png" border="0" alt="[UQ logo]" hspace="20"/>
          </a>
        </td>
        <td>
          <a href="http://www.washington.edu">
            <img src="images/UWlogo150p.gif" border="0" alt="[UW logo]" hspace="20"/>
          </a>
        </td>
        <td>
          <a href="http://infopath.ucsd.edu">
            <img src="images/ucsd_logo.png" border="0" alt="[UCSD logo]" hspace="20"/>
          </a>
        </td>
        <td>
          <a href="http://www.sdsc.edu">
            <img src="images/sdsc_logo.png" border="0" alt="[SDSC logo]" hspace="20"/>
          </a>
        </td>
      </tr>
    </table>
    <table align="center">
      <tr>
        <td>
          <a href="http://nbcr.net">
            <img src="images/nbcr_logo_trans.gif" border="0" alt="[NBCR logo]" hspace="20"/>
          </a>
        </td>
        <td>
          <a href="http://www.cbrc.jp">
            <img src="images/cbrc_logo.png" border="0" alt="[CBRC logo]" hspace="20"/>
          </a>
        </td>
        <td>
          <a href="http://www.ncrr.nih.gov">
            <img src="images/ncrr_logo.gif" border="0" alt="[NCRR logo]" hspace="20"/>
          </a>
        </td>
      </tr>
    </table>
  </body>
</html>

