{
 "cells": [
  {
   "cell_type": "heading",
   "metadata": {},
   "level": 1,
   "source": [
    "                                        HAYSTACK PIPELINE: SELECTION OF HOTSPOTS OF VARIABILITY AND ENRICHED MOTIFS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import os\n",
    "import sys\n",
    "import subprocess as sb\n",
    "import glob\n",
    "import shutil\n",
    "import logging\n",
    "import multiprocessing\n",
    "import platform\n",
    "\n",
    "try:\n",
    "    import cPickle as cp\n",
    "except:\n",
    "    import pickle as cp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "logging.basicConfig(level=logging.INFO,\n",
    "                    format='%(levelname)-5s @ %(asctime)s:\\n\\t %(message)s \\n',\n",
    "                    datefmt='%a, %d %b %Y %H:%M:%S',\n",
    "                    stream=sys.stderr,\n",
    "                    filemode=\"w\", filename='example.log'\n",
    "                    )\n",
    "\n",
    "error = logging.critical\n",
    "warn = logging.warning\n",
    "debug = logging.debug\n",
    "info = logging.info\n",
    "logger = logging.getLogger()\n",
    "logger.setLevel(logging.DEBUG)\n",
    "logging.debug(\"test\")\n",
    "\n",
    "__version__ = \"0.4.0\"\n",
    "HAYSTACK_VERSION = __version__\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ROOT=os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def check_library(library_name):\n",
    "    try:\n",
    "        return __import__(library_name)\n",
    "    except:\n",
    "        error('You need to install %s module to use CRISPResso!' % library_name)\n",
    "        sys.exit(1)\n",
    "\n",
    "def which(program):\n",
    "    import os\n",
    "    def is_exe(fpath):\n",
    "        return os.path.isfile(fpath) and os.access(fpath, os.X_OK)\n",
    "\n",
    "    fpath, fname = os.path.split(program)\n",
    "    if fpath:\n",
    "        if is_exe(program):\n",
    "            return program\n",
    "    else:\n",
    "        for path in os.environ[\"PATH\"].split(os.pathsep):\n",
    "            path = path.strip('\"')\n",
    "            exe_file = os.path.join(path, program)\n",
    "            if is_exe(exe_file):\n",
    "                return exe_file\n",
    "\n",
    "    return None\n",
    "\n",
    "def check_program(binary_name, download_url=None):\n",
    "    if not which(binary_name):\n",
    "        error(\n",
    "            'You need to install and have the command #####%s##### in your PATH variable to use CRISPResso!\\n Please read the documentation!' % binary_name)\n",
    "        if download_url:\n",
    "            error('You can download it from here:%s' % download_url)\n",
    "        sys.exit(1)\n",
    "\n",
    "def check_file(filename):\n",
    "    try:\n",
    "        with open(filename):\n",
    "            pass\n",
    "    except IOError:\n",
    "        raise Exception('I cannot open the file: ' + filename)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "#mandatory\n",
    "parser = argparse.ArgumentParser(description='HAYSTACK Parameters')\n",
    "parser.add_argument('samples_filename_or_bam_folder', type=str,  help='A tab delimeted file with in each row (1) a sample name, (2) the path to the corresponding bam filename, (3 optional) the path to the corresponding gene expression filaneme. Alternatively it is possible to specify a folder containing some .bam files to analyze.')\n",
    "parser.add_argument('genome_name', type=str,  help='Genome assembly to use from UCSC (for example hg19, mm9, etc.)')\n",
    "\n",
    "#optional\n",
    "parser.add_argument('--name',  help='Define a custom output filename for the report', default='')\n",
    "parser.add_argument('--output_directory',type=str, help='Output directory (default: current directory)',default='')\n",
    "parser.add_argument('--bin_size', type=int,help='bin size to use(default: 500bp)',default=500)\n",
    "parser.add_argument('--recompute_all',help='Ignore any file previously precalculated fot the command haystack_hotstpot',action='store_true')\n",
    "parser.add_argument('--depleted', help='Look for cell type specific regions with depletion of signal instead of enrichment',action='store_true')\n",
    "parser.add_argument('--input_is_bigwig', help='Use the bigwig format instead of the bam format for the input. Note: The files must have extension .bw',action='store_true')\n",
    "parser.add_argument('--disable_quantile_normalization',help='Disable quantile normalization (default: False)',action='store_true')\n",
    "parser.add_argument('--transformation',type=str,help='Variance stabilizing transformation among: none, log2, angle (default: angle)',default='angle',choices=['angle', 'log2', 'none'])\n",
    "parser.add_argument('--z_score_high', type=float,help='z-score value to select the specific regions(default: 1.5)',default=1.5)\n",
    "parser.add_argument('--z_score_low', type=float,help='z-score value to select the not specific regions(default: 0.25)',default=0.25)\n",
    "parser.add_argument('--th_rpm',type=float,help='Percentile on the signal intensity to consider for the hotspots (default: 99)', default=99)\n",
    "parser.add_argument('--meme_motifs_filename', type=str, help='Motifs database in MEME format (default JASPAR CORE 2016)')\n",
    "parser.add_argument('--motif_mapping_filename', type=str, help='Custom motif to gene mapping file (the default is for JASPAR CORE 2016 database)')\n",
    "parser.add_argument('--plot_all',  help='Disable the filter on the TF activity and correlation (default z-score TF>0 and rho>0.3)',action='store_true')\n",
    "parser.add_argument('--n_processes',type=int, help='Specify the number of processes to use. The default is #cores available.',default=multiprocessing.cpu_count())\n",
    "parser.add_argument('--temp_directory',  help='Directory to store temporary files  (default: /tmp)', default='/tmp')\n",
    "parser.add_argument('--version',help='Print version and exit.',action='version', version='Version %s' % HAYSTACK_VERSION)\n",
    "args = parser.parse_args()\n",
    "'''\n",
    "SAMPLES_METADATA=os.path.join(_ROOT, 'haystack_analysis/sample_names.txt')\n",
    "OUTPUT_DIR=os.path.join(_ROOT, 'haystack_analysis/output')\n",
    "MOTIF_DIR= os.path.join(_ROOT,'motif_databases')\n",
    "ANNOTATION_DIR=os.path.join(_ROOT,'gene_annotations')\n",
    "GENOME_DIR=os.path.join(_ROOT,'genomes')\n",
    "\n",
    "# HAYSTACK Parameters\n",
    "samples_filename_or_bam_folder = SAMPLES_METADATA\n",
    "genome_name = 'hg19'\n",
    "\n",
    "# optional\n",
    "name = ''\n",
    "output_directory = OUTPUT_DIR\n",
    "bin_size = 200\n",
    "recompute_all = True\n",
    "depleted = True\n",
    "input_is_bigwig = True\n",
    "disable_quantile_normalization = False\n",
    "transformation = 'angle'\n",
    "z_score_high = 1.5\n",
    "z_score_low = 0.25\n",
    "th_rpm = 99\n",
    "meme_motifs_filename = os.path.join(MOTIF_DIR, 'JASPAR_CORE_2016_vertebrates.meme')\n",
    "motif_mapping_filename = os.path.join(MOTIF_DIR, 'JASPAR_CORE_2016_vertebrates_mapped_to_gene_human_mouse.txt')\n",
    "plot_all = True\n",
    "n_processes = multiprocessing.cpu_count()\n",
    "#temp_directory =os.path.join(_ROOT, 'tmp')\n",
    "version = 'Version %s' % HAYSTACK_VERSION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "ename": "Exception",
     "evalue": "I cannot open the file: JASPAR CORE 2016",
     "traceback": [
      "\u001b[0;31m\u001b[0m",
      "\u001b[0;31mException\u001b[0mTraceback (most recent call last)",
      "\u001b[0;32m<ipython-input-36-4d4c0106b538>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mmeme_motifs_filename\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mcheck_file\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmeme_motifs_filename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mmotif_mapping_filename\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mcheck_file\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmotif_mapping_filename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-24-1b645475e0f4>\u001b[0m in \u001b[0;36mcheck_file\u001b[0;34m(filename)\u001b[0m\n\u001b[1;32m     44\u001b[0m             \u001b[0;32mpass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mIOError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 46\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'I cannot open the file: '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mfilename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     47\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mquery_yes_no\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mquestion\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdefault\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"yes\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mException\u001b[0m: I cannot open the file: JASPAR CORE 2016"
     ],
     "output_type": "error"
    }
   ],
   "source": [
    "if meme_motifs_filename:\n",
    "    check_file(meme_motifs_filename)\n",
    "\n",
    "if motif_mapping_filename:\n",
    "    check_file(motif_mapping_filename)\n",
    "    \n",
    "#if not os.path.exists(temp_directory):\n",
    "#    error('The folder specified with --temp_directory: %s does not exist!' % temp_directory)\n",
    " #   sys.exit(1)\n",
    "\n",
    "if input_is_bigwig:\n",
    "        extension_to_check='.bw'\n",
    "        info('Input is set BigWig (.bw)')\n",
    "else:\n",
    "        extension_to_check='.bam'\n",
    "        info('Input is set compressed SAM (.bam)')\n",
    "\n",
    "if name:\n",
    "        directory_name='HAYSTACK_PIPELINE_RESULTS_on_%s' % name\n",
    "\n",
    "else:\n",
    "        directory_name='HAYSTACK_PIPELINE_RESULTS'\n",
    "\n",
    "if output_directory:\n",
    "        output_directory=os.path.join(output_directory,directory_name)\n",
    "else:\n",
    "        output_directory=directory_name\n",
    "\n",
    "#check folder or sample filename\n",
    "\n",
    "USE_GENE_EXPRESSION=True\n",
    "\n",
    "if os.path.isfile(samples_filename_or_bam_folder):\n",
    "        BAM_FOLDER=False\n",
    "        bam_filenames=[]\n",
    "        gene_expression_filenames=[]\n",
    "        sample_names=[]\n",
    "\n",
    "        with open(samples_filename_or_bam_folder) as infile:\n",
    "            for line in infile:\n",
    "\n",
    "                if not line.strip():\n",
    "                        continue\n",
    "                \n",
    "                if line.startswith('#'): #skip optional header line or empty lines\n",
    "                        info('Skipping header/comment line:%s' % line)\n",
    "                        continue\n",
    "\n",
    "                fields=line.strip().split(\"\\t\")\n",
    "                n_fields=len(fields)\n",
    "\n",
    "                if n_fields==2:\n",
    "\n",
    "                    USE_GENE_EXPRESSION=False\n",
    "                    \n",
    "                    sample_names.append(fields[0])\n",
    "                    bam_filenames.append(fields[1])\n",
    "\n",
    "                elif n_fields==3:\n",
    "\n",
    "                    USE_GENE_EXPRESSION=USE_GENE_EXPRESSION and True\n",
    "\n",
    "                    sample_names.append(fields[0])\n",
    "                    bam_filenames.append(fields[1])\n",
    "                    gene_expression_filenames.append(fields[2])\n",
    "                else:\n",
    "                    error('The samples file format is wrong!')\n",
    "        \n",
    "else:\n",
    "        if os.path.exists(samples_filename_or_bam_folder):\n",
    "                BAM_FOLDER=True\n",
    "                USE_GENE_EXPRESSION=False\n",
    "                bam_filenames=glob.glob(os.path.join(samples_filename_or_bam_folder,'*'+extension_to_check))\n",
    "\n",
    "                if not bam_filenames:\n",
    "                    error('No bam/bigwig  files to analyze in %s. Exiting.' % samples_filename_or_bam_folder)\n",
    "                    sys.exit(1)\n",
    "                \n",
    "                sample_names=[os.path.basename(bam_filename).replace(extension_to_check,'') for bam_filename in bam_filenames]\n",
    "        else:\n",
    "                error(\"The file or folder %s doesn't exist. Exiting.\" % samples_filename_or_bam_folder)\n",
    "                sys.exit(1)\n",
    "\n",
    "\n",
    "#check all the files before starting\n",
    "info('Checking samples files location...')\n",
    "for bam_filename in bam_filenames:\n",
    "        check_file(bam_filename)\n",
    "\n",
    "if USE_GENE_EXPRESSION:\n",
    "    for gene_expression_filename in gene_expression_filenames:\n",
    "            check_file(gene_expression_filename)\n",
    "\n",
    "if not os.path.exists(output_directory):\n",
    "        os.makedirs(output_directory)\n",
    "\n",
    "#copy back the file used\n",
    "if not BAM_FOLDER:\n",
    "        shutil.copy2(samples_filename_or_bam_folder,output_directory)\n",
    "\n",
    "#write hotspots conf files\n",
    "sample_names_hotspots_filename=os.path.join(output_directory,'sample_names_hotspots.txt')\n",
    "\n",
    "with open(sample_names_hotspots_filename,'w+') as outfile:\n",
    "    for sample_name,bam_filename in zip(sample_names,bam_filenames):\n",
    "        outfile.write('%s\\t%s\\n' % (sample_name, bam_filename))\n",
    "\n",
    "#write tf activity  conf files\n",
    "if USE_GENE_EXPRESSION:\n",
    "        sample_names_tf_activity_filename=os.path.join(output_directory,'sample_names_tf_activity.txt')\n",
    "\n",
    "        with open(sample_names_tf_activity_filename,'w+') as outfile:\n",
    "                for sample_name,gene_expression_filename in zip(sample_names,gene_expression_filenames):\n",
    "                        outfile.write('%s\\t%s\\n' % (sample_name, gene_expression_filename))\n",
    "\n",
    "        tf_activity_directory=os.path.join(output_directory,'HAYSTACK_TFs_ACTIVITY_PLANES')\n",
    "\n",
    "\n",
    "#CALL HAYSTACK HOTSPOTS\n",
    "cmd_to_run='haystack_hotspots \"%s\" %s --output_directory \"%s\" --bin_size %d %s %s %s %s %s %s %s %s' % \\\n",
    "            (sample_names_hotspots_filename, genome_name,output_directory,bin_size,\n",
    "             ('--recompute_all' if recompute_all else ''),\n",
    "             ('--depleted' if depleted else ''),\n",
    "             ('--input_is_bigwig' if input_is_bigwig else ''),\n",
    "             ('--disable_quantile_normalization' if disable_quantile_normalization else ''),\n",
    "             '--transformation %s' % transformation,\n",
    "             '--z_score_high %f' % z_score_high,\n",
    "             '--z_score_low %f' % z_score_low,\n",
    "             '--th_rpm %f' % th_rpm)\n",
    "print cmd_to_run\n",
    "sb.call(cmd_to_run ,shell=True)        \n",
    "\n",
    "#CALL HAYSTACK MOTIFS\n",
    "motif_directory=os.path.join(output_directory,'HAYSTACK_MOTIFS')\n",
    "for sample_name in sample_names:\n",
    "    specific_regions_filename=os.path.join(output_directory,'HAYSTACK_HOTSPOTS','SPECIFIC_REGIONS','Regions_specific_for_%s*.bed' %sample_name)\n",
    "    bg_regions_filename=glob.glob(os.path.join(output_directory,'HAYSTACK_HOTSPOTS','SPECIFIC_REGIONS','Background_for_%s*.bed' %sample_name))[0]\n",
    "    #bg_regions_filename=glob.glob(specific_regions_filename.replace('Regions_specific','Background')[:-11]+'*.bed')[0] #lo zscore e' diverso...\n",
    "    #print specific_regions_filename,bg_regions_filename\n",
    "    cmd_to_run='haystack_motifs %s %s --bed_bg_filename %s --output_directory %s --name %s' % (specific_regions_filename,genome_name, bg_regions_filename,motif_directory, sample_name)\n",
    "    \n",
    "    if meme_motifs_filename:\n",
    "         cmd_to_run+=' --meme_motifs_filename %s' % meme_motifs_filename\n",
    "         \n",
    "         \n",
    "    if n_processes:\n",
    "        cmd_to_run+=' --n_processes %d' % n_processes\n",
    "        \n",
    "    if temp_directory:\n",
    "        cmd_to_run+=' --temp_directory %s' % temp_directory\n",
    "        \n",
    "        \n",
    "    \n",
    "    print cmd_to_run\n",
    "    sb.call(cmd_to_run,shell=True,env=system_env)\n",
    "\n",
    "    if USE_GENE_EXPRESSION:\n",
    "            #CALL HAYSTACK TF ACTIVITY \n",
    "            motifs_output_folder=os.path.join(motif_directory,'HAYSTACK_MOTIFS_on_%s' % sample_name) \n",
    "            if os.path.exists(motifs_output_folder):\n",
    "                cmd_to_run='haystack_tf_activity_plane %s %s %s --output_directory %s'  %(motifs_output_folder,sample_names_tf_activity_filename,sample_name,tf_activity_directory)\n",
    "                \n",
    "                if motif_mapping_filename:\n",
    "                    cmd_to_run+=' --motif_mapping_filename %s' %  motif_mapping_filename       \n",
    "                \n",
    "                if plot_all:\n",
    "                    cmd_to_run+=' --plot_all'\n",
    "                    \n",
    "                \n",
    "                print cmd_to_run\n",
    "                sb.call(cmd_to_run,shell=True,env=system_env)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/hd2/Dropbox (Partners HealthCare)/PROJECTS/2017_07_HAYSTACK/haystack_analysis/sample_names.txt'"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    ""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2.0
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}