{
 "cells": [
  {
   "cell_type": "heading",
   "metadata": {},
   "level": 1,
   "source": [
    "                                        HAYSTACK PIPELINE: SELECTION OF HOTSPOTS OF VARIABILITY AND ENRICHED MOTIFS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import os\n",
    "import sys\n",
    "import subprocess as sb\n",
    "import glob\n",
    "import shutil\n",
    "import multiprocessing\n",
    "\n",
    "try:\n",
    "    import cPickle as cp\n",
    "except:\n",
    "    import pickle as cp\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "\n",
    "mpl.use('Agg')\n",
    "import pylab as pl\n",
    "import xml.etree.cElementTree as ET\n",
    "from pybedtools import BedTool\n",
    "#from bioutilities import Genome_2bit\n",
    "import pysam\n",
    "import re\n",
    "import numpy as np\n",
    "from scipy.stats import zscore\n",
    "import urllib2\n",
    "\n",
    "HAYSTACK_VERSION = \"0.4.0\"\n",
    "\n",
    "import logging\n",
    "logging.basicConfig(level=logging.DEBUG,\n",
    "                    format='%(levelname)-5s @ %(asctime)s:\\n\\t %(message)s \\n',\n",
    "                    datefmt='%a, %d %b %Y %H:%M:%S',\n",
    "                    stream=sys.stderr,\n",
    "                    filemode=\"w\", filename='example.log'\n",
    "                    )\n",
    "\n",
    "error = logging.critical\n",
    "warn = logging.warning\n",
    "debug = logging.debug\n",
    "info = logging.info\n",
    "logger = logging.getLogger()\n",
    "logger.setLevel(logging.DEBUG)\n",
    "logging.debug(\"start\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# commmon functions haystack hotspots\n",
    "def check_library(library_name):\n",
    "    try:\n",
    "        return __import__(library_name)\n",
    "    except:\n",
    "        error('You need to install %s module to use haystack!' % library_name)\n",
    "        sys.exit(1)\n",
    "\n",
    "def which(program):\n",
    "    def is_exe(fpath):\n",
    "        return os.path.isfile(fpath) and os.access(fpath, os.X_OK)\n",
    "\n",
    "    fpath, fname = os.path.split(program)\n",
    "    if fpath:\n",
    "        if is_exe(program):\n",
    "            return program\n",
    "    else:\n",
    "        for path in os.environ[\"PATH\"].split(os.pathsep):\n",
    "            path = path.strip('\"')\n",
    "            exe_file = os.path.join(path, program)\n",
    "            if is_exe(exe_file):\n",
    "                return exe_file\n",
    "    return None\n",
    "\n",
    "def check_program(binary_name, download_url=None):\n",
    "    if not which(binary_name):\n",
    "        error(\n",
    "            'You need to install and have the command #####%s##### in your PATH variable to use CRISPResso!\\n Please read the documentation!' % binary_name)\n",
    "        if download_url:\n",
    "            error('You can download it from here:%s' % download_url)\n",
    "        sys.exit(1)\n",
    "\n",
    "def check_file(filename):\n",
    "    try:\n",
    "        with open(filename):\n",
    "            pass\n",
    "    except IOError:\n",
    "        raise Exception('I cannot open the file: ' + filename)\n",
    "\n",
    "def quantile_normalization(A):\n",
    "    AA = np.zeros_like(A)\n",
    "    I = np.argsort(A, axis=0)\n",
    "    AA[I, np.arange(A.shape[1])] = np.mean(A[I, np.arange(A.shape[1])], axis=1)[:, np.newaxis]\n",
    "\n",
    "    return AA\n",
    "\n",
    "def smooth(x, window_len=200):\n",
    "    s = np.r_[x[window_len - 1:0:-1], x, x[-1:-window_len:-1]]\n",
    "    w = np.hanning(window_len)\n",
    "    y = np.convolve(w / w.sum(), s, mode='valid')\n",
    "    return y[int(window_len / 2):-int(window_len / 2) + 1]\n",
    "\n",
    "# write the IGV session file\n",
    "def rem_base_path(path, base_path):\n",
    "    return path.replace(os.path.join(base_path, ''), '')\n",
    "\n",
    "def find_th_rpm(df_chip, th_rpm_param):\n",
    "    return np.min(df_chip.apply(lambda x: np.percentile(x, th_rpm_param)))\n",
    "\n",
    "def log2_transform(x):\n",
    "    return np.log2(x + 1)\n",
    "\n",
    "def angle_transform(x):\n",
    "    return np.arcsin(np.sqrt(x) / 1000000.0)\n",
    "\n",
    "def download_genome(name, output_directory=None):\n",
    "    urlpath = \"http://hgdownload.cse.ucsc.edu/goldenPath/%s/bigZips/%s.2bit\" % (name, name)\n",
    "    genome_url_origin = urllib2.urlopen(urlpath)\n",
    "\n",
    "    genome_filename = os.path.join(output_directory, \"%s.2bit\" % name)\n",
    "\n",
    "    print 'Downloading %s in %s...' % (urlpath, genome_filename)\n",
    "\n",
    "    if os.path.exists(genome_filename):\n",
    "        print 'File %s exists, skipping download' % genome_filename\n",
    "    else:\n",
    "\n",
    "        with open(genome_filename, 'wb') as genome_file_destination:\n",
    "            shutil.copyfileobj(genome_url_origin, genome_file_destination)\n",
    "\n",
    "        print 'Downloded %s in %s:' % (urlpath, genome_filename)\n",
    "\n",
    "    g = Genome_2bit(genome_filename, verbose=True)\n",
    "\n",
    "    chr_len_filename = os.path.join(output_directory, \"%s_chr_lengths.txt\" % name)\n",
    "    if not os.path.exists(chr_len_filename):\n",
    "        print 'Extracting chromosome lengths'\n",
    "        g.write_chr_len(chr_len_filename)\n",
    "        print 'Done!'\n",
    "    else:\n",
    "        print 'File %s exists, skipping generation' % chr_len_filename\n",
    "\n",
    "    meme_bg_filename = os.path.join(output_directory, \"%s_meme_bg\" % name)\n",
    "    if not os.path.exists(meme_bg_filename):\n",
    "        print 'Calculating nucleotide frequencies....'\n",
    "        g.write_meme_background(meme_bg_filename)\n",
    "        print 'Done!'\n",
    "    else:\n",
    "        print 'File %s exists, skipping generation' % meme_bg_filename\n",
    "\n",
    "def determine_path(folder):\n",
    "    _ROOT = os.getcwd()\n",
    "    return os.path.join(_ROOT, folder)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determine pipeline folder paths"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_filename = determine_path('test_data/samples_names.txt')\n",
    "output_directory = determine_path('haystack_analysis/output')\n",
    "motif_directory = determine_path('motif_databases')\n",
    "annotation_directory = determine_path('gene_annotations')\n",
    "genome_directory = determine_path('genomes')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_filename_or_bam_folder = samples_filename\n",
    "genome_name = 'hg19'\n",
    "# optional\n",
    "name = ''\n",
    "output_directory = output_directory\n",
    "bin_size = 200\n",
    "recompute_all = True\n",
    "depleted = True\n",
    "input_is_bigwig = False\n",
    "disable_quantile_normalization = False\n",
    "transformation = 'angle'\n",
    "z_score_high = 1.5\n",
    "z_score_low = 0.25\n",
    "th_rpm = 99\n",
    "meme_motifs_filename = ''  # os.path.join(MOTIF_DIR, 'JASPAR_CORE_2016_vertebrates.meme')\n",
    "motif_mapping_filename = ''  # os.path.join(MOTIF_DIR, 'JASPAR_CORE_2016_vertebrates_mapped_to_gene_human_mouse.txt')\n",
    "plot_all = True\n",
    "use_X_Y = True\n",
    "max_regions_percentage=0.1\n",
    "n_processes = multiprocessing.cpu_count()\n",
    "temp_directory = ''  # os.path.join(_ROOT, 'tmp')\n",
    "version = 'Version %s' % HAYSTACK_VERSION\n",
    "# add two flags\n",
    "chrom_exclude = ''  # '_|chrM|chrX|chrY'\n",
    "ext = 200 # read extension in bps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Intialize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "if depleted:\n",
    "    z_score_high = -z_score_high\n",
    "    z_score_low = -z_score_low\n",
    "\n",
    "if meme_motifs_filename:\n",
    "    check_file(meme_motifs_filename)\n",
    "\n",
    "if motif_mapping_filename:\n",
    "    check_file(motif_mapping_filename)\n",
    "\n",
    "# if not os.path.exists(temp_directory):\n",
    "#    error('The folder specified with --temp_directory: %s does not exist!' % temp_directory)\n",
    "#    sys.exit(1)\n",
    "\n",
    "if input_is_bigwig:\n",
    "    extension_to_check = '.bw'\n",
    "    info('Input is set BigWig (.bw)')\n",
    "else:\n",
    "    extension_to_check = '.bam'\n",
    "    info('Input is set compressed SAM (.bam)')\n",
    "\n",
    "if name:\n",
    "    directory_name = 'HAYSTACK_PIPELINE_RESULTS_on_%s' % name\n",
    "\n",
    "else:\n",
    "    directory_name = 'HAYSTACK_PIPELINE_RESULTS'\n",
    "\n",
    "if output_directory:\n",
    "    output_directory = os.path.join(output_directory, directory_name)\n",
    "else:\n",
    "    output_directory = directory_name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "read data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check folder or sample filename\n",
    "\n",
    "USE_GENE_EXPRESSION = True\n",
    "\n",
    "if os.path.isfile(samples_filename_or_bam_folder):\n",
    "    BAM_FOLDER = False\n",
    "    bam_filenames = []\n",
    "    gene_expression_filenames = []\n",
    "    sample_names = []\n",
    "\n",
    "    dir_path = os.path.dirname(os.path.realpath(samples_filename_or_bam_folder))\n",
    "\n",
    "    with open(samples_filename_or_bam_folder) as infile:\n",
    "        for line in infile:\n",
    "\n",
    "            if not line.strip():\n",
    "                continue\n",
    "\n",
    "            if line.startswith('#'):  # skip optional header line or empty lines\n",
    "                info('Skipping header/comment line:%s' % line)\n",
    "                continue\n",
    "\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            n_fields = len(fields)\n",
    "\n",
    "            if n_fields == 2:\n",
    "\n",
    "                USE_GENE_EXPRESSION = False\n",
    "\n",
    "                sample_names.append(fields[0])\n",
    "                bam_filenames.append(fields[1])\n",
    "\n",
    "            elif n_fields == 3:\n",
    "\n",
    "                USE_GENE_EXPRESSION = USE_GENE_EXPRESSION and True\n",
    "\n",
    "                sample_names.append(fields[0])\n",
    "                bam_filenames.append(fields[1])\n",
    "                gene_expression_filenames.append(fields[2])\n",
    "            else:\n",
    "                error('The samples file format is wrong!')\n",
    "\n",
    "    bam_filenames = [os.path.join(dir_path, filename) for filename in bam_filenames]\n",
    "    gene_expression_filenames = [os.path.join(dir_path, filename) for filename in gene_expression_filenames]\n",
    "\n",
    "else:\n",
    "    if os.path.exists(samples_filename_or_bam_folder):\n",
    "        BAM_FOLDER = True\n",
    "        USE_GENE_EXPRESSION = False\n",
    "        bam_filenames = glob.glob(os.path.join(samples_filename_or_bam_folder, '*' + extension_to_check))\n",
    "\n",
    "        if not bam_filenames:\n",
    "            error('No bam/bigwig  files to analyze in %s. Exiting.' % samples_filename_or_bam_folder)\n",
    "            sys.exit(1)\n",
    "\n",
    "        sample_names = [os.path.basename(bam_filename).replace(extension_to_check, '') for bam_filename in\n",
    "                        bam_filenames]\n",
    "    else:\n",
    "        error(\"The file or folder %s doesn't exist. Exiting.\" % samples_filename_or_bam_folder)\n",
    "        sys.exit(1)\n",
    "\n",
    "# check all the files before starting\n",
    "info('Checking samples files location...')\n",
    "for bam_filename in bam_filenames:\n",
    "    check_file(bam_filename)\n",
    "\n",
    "if USE_GENE_EXPRESSION:\n",
    "    for gene_expression_filename in gene_expression_filenames:\n",
    "        check_file(gene_expression_filename)\n",
    "\n",
    "if not os.path.exists(output_directory):\n",
    "    os.makedirs(output_directory)\n",
    "\n",
    "# copy back the file used\n",
    "if not BAM_FOLDER:\n",
    "    shutil.copy2(samples_filename_or_bam_folder, output_directory)\n",
    "\n",
    "# write hotspots conf files\n",
    "sample_names_hotspots_filename = os.path.join(output_directory, 'sample_names_hotspots.txt')\n",
    "\n",
    "with open(sample_names_hotspots_filename, 'w+') as outfile:\n",
    "    for sample_name, bam_filename in zip(sample_names, bam_filenames):\n",
    "        outfile.write('%s\\t%s\\n' % (sample_name, bam_filename))\n",
    "\n",
    "# write tf activity  conf files\n",
    "if USE_GENE_EXPRESSION:\n",
    "    sample_names_tf_activity_filename = os.path.join(output_directory, 'sample_names_tf_activity.txt')\n",
    "\n",
    "    with open(sample_names_tf_activity_filename, 'w+') as outfile:\n",
    "        for sample_name, gene_expression_filename in zip(sample_names, gene_expression_filenames):\n",
    "            outfile.write('%s\\t%s\\n' % (sample_name, gene_expression_filename))\n",
    "\n",
    "    tf_activity_directory = os.path.join(output_directory, 'HAYSTACK_TFs_ACTIVITY_PLANES')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "create HAYSTACK_HOTSPOTS directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "if name:\n",
    "    directory_name = 'HAYSTACK_HOTSPOTS_on_%s' % name\n",
    "\n",
    "else:\n",
    "    directory_name = 'HAYSTACK_HOTSPOTS'\n",
    "\n",
    "if output_directory:\n",
    "    output_directory = os.path.join(output_directory, directory_name)\n",
    "else:\n",
    "    output_directory = directory_name\n",
    "\n",
    "if not os.path.exists(output_directory):\n",
    "    os.makedirs(output_directory)\n",
    "\n",
    "tracks_directory = os.path.join(output_directory, 'TRACKS')\n",
    "if not os.path.exists(tracks_directory):\n",
    "    os.makedirs(tracks_directory)\n",
    "\n",
    "intermediate_directory = os.path.join(output_directory, 'INTERMEDIATE')\n",
    "if not os.path.exists(intermediate_directory):\n",
    "    os.makedirs(intermediate_directory)\n",
    "\n",
    "\n",
    "specific_regions_directory = os.path.join(output_directory, 'SPECIFIC_REGIONS')\n",
    "if not os.path.exists(specific_regions_directory):\n",
    "        os.makedirs(specific_regions_directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create filepaths to required files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input files\n",
    "chr_len_filename = os.path.join(genome_directory, \"%s_chr_lengths.txt\" % genome_name)\n",
    "blacklist = os.path.join(genome_directory, 'blacklist.bed')\n",
    "check_file(chr_len_filename)\n",
    "check_file(blacklist)\n",
    "\n",
    "# output files\n",
    "genome_sorted_filtered_bins_file = os.path.join(output_directory,\n",
    "                                                '%s.%dbp.bins.sorted.filterd.bed' % (os.path.basename(genome_name),\n",
    "                                                                                     bin_size))\n",
    "bedgraph_iod_track_filename = os.path.join(tracks_directory,\n",
    "                                           'VARIABILITY.bedgraph')\n",
    "bw_iod_track_filename = os.path.join(tracks_directory,\n",
    "                                     'VARIABILITY.bw')\n",
    "bedgraph_hpr_filename = os.path.join(tracks_directory,\n",
    "                                     'SELECTED_VARIABILITY_HOTSPOT.bedgraph')\n",
    "bed_hpr_fileaname = os.path.join(output_directory,\n",
    "                                 'SELECTED_VARIABILITY_HOTSPOT.bed')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Required functions for the pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create filepaths to required files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# step 1\n",
    "info('Initializing Genome:%s' % genome_name)\n",
    "intialize_genome(genome_name)\n",
    "\n",
    "# step 2\n",
    "info('Creating bins of %dbp in %s' % (bin_size, genome_sorted_filtered_bins_file))\n",
    "if not os.path.exists(genome_sorted_filtered_bins_file) or recompute_all:\n",
    "    create_tiled_genome(genome_sorted_filtered_bins_file)\n",
    "\n",
    "# step 3\n",
    "info('Convert files to genome-wide rpm tracks')\n",
    "to_rpm_tracks(sample_names, bam_filenames, skipfilter=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# step 4\n",
    "info('Normalize rpm tracks')\n",
    "binned_rpm_filenames = glob.glob(os.path.join(intermediate_directory, '*.rpm'))\n",
    "df_chip, coordinates_bin =load_rpm_tracks(binned_rpm_filenames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GM12878.200bp</th>\n",
       "      <th>H1hesc.200bp</th>\n",
       "      <th>HEPG2.200bp</th>\n",
       "      <th>HSMM.200bp</th>\n",
       "      <th>K562.200bp</th>\n",
       "      <th>NHLF.200bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.038836</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.069769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.077672</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.068074</td>\n",
       "      <td>0.069769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GM12878.200bp</th>\n",
       "      <th>H1hesc.200bp</th>\n",
       "      <th>HEPG2.200bp</th>\n",
       "      <th>HSMM.200bp</th>\n",
       "      <th>K562.200bp</th>\n",
       "      <th>NHLF.200bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.038836</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.069769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.077672</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.068074</td>\n",
       "      <td>0.069769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chip.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>bpstart</th>\n",
       "      <th>bpend</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>10200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10200</td>\n",
       "      <td>10400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11600</td>\n",
       "      <td>11800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11800</td>\n",
       "      <td>12000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12000</td>\n",
       "      <td>12200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12200</td>\n",
       "      <td>12400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12400</td>\n",
       "      <td>12600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12600</td>\n",
       "      <td>12800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12800</td>\n",
       "      <td>13000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13000</td>\n",
       "      <td>13200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>bpstart</th>\n",
       "      <th>bpend</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>10200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10200</td>\n",
       "      <td>10400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11600</td>\n",
       "      <td>11800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11800</td>\n",
       "      <td>12000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12000</td>\n",
       "      <td>12200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12200</td>\n",
       "      <td>12400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12400</td>\n",
       "      <td>12600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12600</td>\n",
       "      <td>12800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12800</td>\n",
       "      <td>13000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13000</td>\n",
       "      <td>13200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordinates_bin.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "if disable_quantile_normalization:\n",
    "    info('Skipping quantile normalization...')\n",
    "    df_chip_to_bigwig(df_chip, coordinates_bin)\n",
    "\n",
    "else:\n",
    "    info('Normalizing the data...')\n",
    "    df_chip = pd.DataFrame(quantile_normalization(df_chip.values),\n",
    "                               columns=df_chip.columns,\n",
    "                               index=df_chip.index)\n",
    "    df_chip_to_bigwig(df_chip, coordinates_bin, normalized=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GM12878.200bp</th>\n",
       "      <th>H1hesc.200bp</th>\n",
       "      <th>HEPG2.200bp</th>\n",
       "      <th>HSMM.200bp</th>\n",
       "      <th>K562.200bp</th>\n",
       "      <th>NHLF.200bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006473</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.071288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.085606</td>\n",
       "      <td>0.101537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GM12878.200bp</th>\n",
       "      <th>H1hesc.200bp</th>\n",
       "      <th>HEPG2.200bp</th>\n",
       "      <th>HSMM.200bp</th>\n",
       "      <th>K562.200bp</th>\n",
       "      <th>NHLF.200bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006473</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.071288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.085606</td>\n",
       "      <td>0.101537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.040222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05185</td>\n",
       "      <td>0.012945</td>\n",
       "      <td>0.028876</td>\n",
       "      <td>0.012945</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chip.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# step 5\n",
    "info('Determine HP regions')\n",
    "hpr_idxs, coordinates_bin, df_chip_hpr_zscore, hpr_iod_scores = find_hpr_coordinates(df_chip, coordinates_bin)\n",
    "info('Save files')\n",
    "hpr_to_bigwig(coordinates_bin)\n",
    "hpr_to_bedgraph(hpr_idxs, coordinates_bin)\n",
    "write_specific_regions(coordinates_bin, df_chip_hpr_zscore)\n",
    "create_igv_track_file(hpr_iod_scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4370    0.000002\n4371    0.000001\n4446    0.000001\n4447    0.000002\n5147    0.000001\n5148    0.000001\n5875    0.000002\n8516    0.000001\n9474    0.000001\n9475    0.000002\ndtype: float64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hpr_iod_scores.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>bpstart</th>\n",
       "      <th>bpend</th>\n",
       "      <th>iod</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2729</th>\n",
       "      <td>chr1</td>\n",
       "      <td>713600</td>\n",
       "      <td>713800</td>\n",
       "      <td>1.522046e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2730</th>\n",
       "      <td>chr1</td>\n",
       "      <td>713800</td>\n",
       "      <td>714000</td>\n",
       "      <td>1.360054e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3540</th>\n",
       "      <td>chr1</td>\n",
       "      <td>879200</td>\n",
       "      <td>879400</td>\n",
       "      <td>5.883203e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3541</th>\n",
       "      <td>chr1</td>\n",
       "      <td>879400</td>\n",
       "      <td>879600</td>\n",
       "      <td>5.085509e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3614</th>\n",
       "      <td>chr1</td>\n",
       "      <td>894000</td>\n",
       "      <td>894200</td>\n",
       "      <td>2.181267e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3615</th>\n",
       "      <td>chr1</td>\n",
       "      <td>894200</td>\n",
       "      <td>894400</td>\n",
       "      <td>1.283248e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3616</th>\n",
       "      <td>chr1</td>\n",
       "      <td>894400</td>\n",
       "      <td>894600</td>\n",
       "      <td>9.571502e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3652</th>\n",
       "      <td>chr1</td>\n",
       "      <td>901600</td>\n",
       "      <td>901800</td>\n",
       "      <td>5.096526e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3653</th>\n",
       "      <td>chr1</td>\n",
       "      <td>901800</td>\n",
       "      <td>902000</td>\n",
       "      <td>6.060991e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3654</th>\n",
       "      <td>chr1</td>\n",
       "      <td>902000</td>\n",
       "      <td>902200</td>\n",
       "      <td>5.704145e-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>bpstart</th>\n",
       "      <th>bpend</th>\n",
       "      <th>iod</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2729</th>\n",
       "      <td>chr1</td>\n",
       "      <td>713600</td>\n",
       "      <td>713800</td>\n",
       "      <td>1.522046e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2730</th>\n",
       "      <td>chr1</td>\n",
       "      <td>713800</td>\n",
       "      <td>714000</td>\n",
       "      <td>1.360054e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3540</th>\n",
       "      <td>chr1</td>\n",
       "      <td>879200</td>\n",
       "      <td>879400</td>\n",
       "      <td>5.883203e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3541</th>\n",
       "      <td>chr1</td>\n",
       "      <td>879400</td>\n",
       "      <td>879600</td>\n",
       "      <td>5.085509e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3614</th>\n",
       "      <td>chr1</td>\n",
       "      <td>894000</td>\n",
       "      <td>894200</td>\n",
       "      <td>2.181267e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3615</th>\n",
       "      <td>chr1</td>\n",
       "      <td>894200</td>\n",
       "      <td>894400</td>\n",
       "      <td>1.283248e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3616</th>\n",
       "      <td>chr1</td>\n",
       "      <td>894400</td>\n",
       "      <td>894600</td>\n",
       "      <td>9.571502e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3652</th>\n",
       "      <td>chr1</td>\n",
       "      <td>901600</td>\n",
       "      <td>901800</td>\n",
       "      <td>5.096526e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3653</th>\n",
       "      <td>chr1</td>\n",
       "      <td>901800</td>\n",
       "      <td>902000</td>\n",
       "      <td>6.060991e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3654</th>\n",
       "      <td>chr1</td>\n",
       "      <td>902000</td>\n",
       "      <td>902200</td>\n",
       "      <td>5.704145e-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordinates_bin.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GM12878.200bp</th>\n",
       "      <th>H1hesc.200bp</th>\n",
       "      <th>HEPG2.200bp</th>\n",
       "      <th>HSMM.200bp</th>\n",
       "      <th>K562.200bp</th>\n",
       "      <th>NHLF.200bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4370</th>\n",
       "      <td>-0.488069</td>\n",
       "      <td>-0.311384</td>\n",
       "      <td>-0.387298</td>\n",
       "      <td>-0.458339</td>\n",
       "      <td>2.228175</td>\n",
       "      <td>-0.583085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4371</th>\n",
       "      <td>-0.459942</td>\n",
       "      <td>-0.478323</td>\n",
       "      <td>-0.272010</td>\n",
       "      <td>-0.574770</td>\n",
       "      <td>2.227086</td>\n",
       "      <td>-0.442040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4446</th>\n",
       "      <td>-0.536726</td>\n",
       "      <td>-0.438270</td>\n",
       "      <td>2.234039</td>\n",
       "      <td>-0.410387</td>\n",
       "      <td>-0.438270</td>\n",
       "      <td>-0.410387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4447</th>\n",
       "      <td>-0.547223</td>\n",
       "      <td>-0.426057</td>\n",
       "      <td>2.233670</td>\n",
       "      <td>-0.426057</td>\n",
       "      <td>-0.408275</td>\n",
       "      <td>-0.426057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5147</th>\n",
       "      <td>-0.994770</td>\n",
       "      <td>-0.151963</td>\n",
       "      <td>-0.871500</td>\n",
       "      <td>1.220835</td>\n",
       "      <td>-0.701180</td>\n",
       "      <td>1.498577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5148</th>\n",
       "      <td>-0.879217</td>\n",
       "      <td>-0.274687</td>\n",
       "      <td>-0.807228</td>\n",
       "      <td>1.081352</td>\n",
       "      <td>-0.773283</td>\n",
       "      <td>1.653064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5875</th>\n",
       "      <td>-0.402657</td>\n",
       "      <td>2.224369</td>\n",
       "      <td>-0.448244</td>\n",
       "      <td>-0.289137</td>\n",
       "      <td>-0.448244</td>\n",
       "      <td>-0.636087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8516</th>\n",
       "      <td>-0.447970</td>\n",
       "      <td>-0.332391</td>\n",
       "      <td>2.226650</td>\n",
       "      <td>-0.556949</td>\n",
       "      <td>-0.332391</td>\n",
       "      <td>-0.556949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9474</th>\n",
       "      <td>-0.673379</td>\n",
       "      <td>-0.512369</td>\n",
       "      <td>2.208026</td>\n",
       "      <td>-0.430424</td>\n",
       "      <td>-0.451251</td>\n",
       "      <td>-0.140602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9475</th>\n",
       "      <td>-0.604159</td>\n",
       "      <td>-0.382655</td>\n",
       "      <td>2.180051</td>\n",
       "      <td>-0.578032</td>\n",
       "      <td>-0.633627</td>\n",
       "      <td>0.018422</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GM12878.200bp</th>\n",
       "      <th>H1hesc.200bp</th>\n",
       "      <th>HEPG2.200bp</th>\n",
       "      <th>HSMM.200bp</th>\n",
       "      <th>K562.200bp</th>\n",
       "      <th>NHLF.200bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4370</th>\n",
       "      <td>-0.488069</td>\n",
       "      <td>-0.311384</td>\n",
       "      <td>-0.387298</td>\n",
       "      <td>-0.458339</td>\n",
       "      <td>2.228175</td>\n",
       "      <td>-0.583085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4371</th>\n",
       "      <td>-0.459942</td>\n",
       "      <td>-0.478323</td>\n",
       "      <td>-0.272010</td>\n",
       "      <td>-0.574770</td>\n",
       "      <td>2.227086</td>\n",
       "      <td>-0.442040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4446</th>\n",
       "      <td>-0.536726</td>\n",
       "      <td>-0.438270</td>\n",
       "      <td>2.234039</td>\n",
       "      <td>-0.410387</td>\n",
       "      <td>-0.438270</td>\n",
       "      <td>-0.410387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4447</th>\n",
       "      <td>-0.547223</td>\n",
       "      <td>-0.426057</td>\n",
       "      <td>2.233670</td>\n",
       "      <td>-0.426057</td>\n",
       "      <td>-0.408275</td>\n",
       "      <td>-0.426057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5147</th>\n",
       "      <td>-0.994770</td>\n",
       "      <td>-0.151963</td>\n",
       "      <td>-0.871500</td>\n",
       "      <td>1.220835</td>\n",
       "      <td>-0.701180</td>\n",
       "      <td>1.498577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5148</th>\n",
       "      <td>-0.879217</td>\n",
       "      <td>-0.274687</td>\n",
       "      <td>-0.807228</td>\n",
       "      <td>1.081352</td>\n",
       "      <td>-0.773283</td>\n",
       "      <td>1.653064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5875</th>\n",
       "      <td>-0.402657</td>\n",
       "      <td>2.224369</td>\n",
       "      <td>-0.448244</td>\n",
       "      <td>-0.289137</td>\n",
       "      <td>-0.448244</td>\n",
       "      <td>-0.636087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8516</th>\n",
       "      <td>-0.447970</td>\n",
       "      <td>-0.332391</td>\n",
       "      <td>2.226650</td>\n",
       "      <td>-0.556949</td>\n",
       "      <td>-0.332391</td>\n",
       "      <td>-0.556949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9474</th>\n",
       "      <td>-0.673379</td>\n",
       "      <td>-0.512369</td>\n",
       "      <td>2.208026</td>\n",
       "      <td>-0.430424</td>\n",
       "      <td>-0.451251</td>\n",
       "      <td>-0.140602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9475</th>\n",
       "      <td>-0.604159</td>\n",
       "      <td>-0.382655</td>\n",
       "      <td>2.180051</td>\n",
       "      <td>-0.578032</td>\n",
       "      <td>-0.633627</td>\n",
       "      <td>0.018422</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chip_hpr_zscore.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "info('All done! Ciao!')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2.0
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}